### ISSUE TITLE
Crash once putting in GetUrlConfig value

### ISSUE THREAD

**Issue by H3jake** [DigitalRuby/IPBan#225]
Hi, I'm having a weird issue with the config, i have removed all of the stuff i dont need from it and that works fine but once i enter a URL for future updates i get the following error:
2022-10-21 12:36:12.9162|WARN|IPBan|Initializing service
2022-10-21 12:36:13.1241|WARN|IPBan|Detecting os version...
2022-10-21 12:36:13.2814|WARN|IPBan|OS version detected: Name: Windows, Version: 10.0.14393, Friendly Name: Microsoft Windows Server 2016 Standard, Description: Microsoft Windows 10.0.14393, app version: 1.8.0
2022-10-21 12:36:13.3392|WARN|IPBan|Running as a console app
2022-10-21 12:36:13.5126|WARN|IPBan|Preparing to run service
2022-10-21 12:36:13.5295|WARN|IPBan|Starting service
2022-10-21 12:36:13.5316|WARN|IPBan|Running service
2022-10-21 12:36:13.6446|WARN|IPBan|IPBan is free software created and refined over many years.
2022-10-21 12:36:13.6446|WARN|IPBan|Please consider upgrading to the pro version for more advanced functions, shared ban lists and much more.
2022-10-21 12:36:13.6446|WARN|IPBan|Learn more at https://ipban.com
2022-10-21 12:36:13.6559|INFO|IPBan|Initializing IPBan database at Data Source=C:\Program Files\IPBan\ipban.sqlite;Cache=Shared;
2022-10-21 12:36:14.0029|WARN|IPBan|IPBan service started and initialized
2022-10-21 12:36:14.0029|WARN|IPBan|Log levels: True,True,True,True,False,False
2022-10-21 12:36:18.5356|INFO|IPBan|Config file changed
2022-10-21 12:36:18.8031|WARN|IPBan|Loaded firewall type DigitalRuby.IPBanCore.IPBanWindowsFirewall
2022-10-21 12:36:18.8305|WARN|IPBan|Clearing all banned ip addresses on start because ClearBannedIPAddressesOnRestart is set
2022-10-21 12:36:18.9430|WARN|IPBan|Initialized event viewer with query string: <QueryList><Query Id='1' Path='Security'><Select Path='Security'>*[System[(band(Keywords,9227875636482146304))]]</Select></Query><Query Id='2' Path='Security'><Select Path='Security'>*[System[(band(Keywords,9227875636482146304))]]</Select></Query></QueryList>
2022-10-21 12:36:18.9430|WARN|IPBan|Ignoring event viewer paths: 
2022-10-21 12:36:19.1054|INFO|IPBan|Local ip address: x.x.x.x
2022-10-21 12:36:20.3224|INFO|IPBan|Remote ip address: y.y.y.y
2022-10-21 12:36:20.3224|INFO|IPBan|FQDN: server.company.com
2022-10-21 12:36:20.3746|ERROR|IPBan|Error getting url of type Config at http://URL/ipban.config: System.Xml.XmlException: Data at the root level is invalid. Line 1, position 1.
   at System.Xml.XmlTextReaderImpl.Throw(Exception )
   at System.Xml.XmlTextReaderImpl.Throw(String , String )
   at System.Xml.XmlTextReaderImpl.Throw(String )
   at System.Xml.XmlTextReaderImpl.ParseRootLevelWhitespace()
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlLoader.Load(XmlDocument , XmlReader , Boolean )
   at System.Xml.XmlDocument.Load(XmlReader )
   at DigitalRuby.IPBanCore.IPBanService.WriteConfigAsync(String xml)
   at DigitalRuby.IPBanCore.IPBanService.GetUrl(UrlType urlType)
2022-10-21 12:36:20.4496|WARN|IPBan|IPBan is running correctly. You can perform a failed ssh or rdp login and watch this log to verify functionality.

**Comment by H3jake** [comment]
Here is the config file:
`
<?xml version="1.0"?>

<configuration>
  <LogFilesToParse>
    <LogFiles>
    </LogFiles>
  </LogFilesToParse>
  <ExpressionsToBlock>
    <Groups>
      <Group>
        <Source>RDP</Source>
        <Keywords>0x8010000000000000</Keywords>
        <Path>Security</Path>
        <FailedLoginThreshold>0</FailedLoginThreshold>		
        <Expressions>
          <Expression>
            <XPath>//EventID</XPath>
            <Regex>
				<![CDATA[
				    ^(?<log>4625|5152)$
			    ]]>
			</Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='IpAddress' or @Name='Workstation' or @Name='SourceAddress']</XPath>
            <Regex>
              <![CDATA[
                (?<ipaddress>.+)
              ]]>
            </Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='ProcessName']</XPath>
            <Regex>
              <![CDATA[
                (?<source_IIS>c:\\Windows\\System32\\inetsrv\\w3wp.exe)?$
              ]]>
            </Regex>
          </Expression>
        </Expressions>
      </Group>
      <Group>
        <Source>RDP</Source>
        <Keywords>0x8010000000000000</Keywords>
        <Path>Security</Path>
        <FailedLoginThreshold>0</FailedLoginThreshold>
        <Expressions>
          <Expression>
            <XPath>//EventID</XPath>
            <Regex>
				<![CDATA[
				    ^(?<log>4653)$
				]]>
			</Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='FailureReason']</XPath>
            <Regex>.</Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='RemoteAddress']</XPath>
            <Regex>
              <![CDATA[
                (?<ipaddress>.+)
              ]]>
            </Regex>
          </Expression>
        </Expressions>
      </Group>        
    </Groups>
  </ExpressionsToBlock>
  <ExpressionsToNotify>
    <Groups>
      <Group>
        <Source>RDP</Source>
        <Keywords>0x8020000000000000</Keywords>
        <Path>Security</Path>
        <Expressions>
          <Expression>
            <XPath>//EventID</XPath>
            <Regex>^4624$</Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='ProcessName' or @Name='LogonProcessName']</XPath>
            <Regex>winlogon|svchost|ntlmssp</Regex>
          </Expression>
          <Expression>
            <XPath>//Data[@Name='IpAddress' or @Name='Workstation' or @Name='SourceAddress']</XPath>
            <Regex>
              <![CDATA[
                (?<ipaddress>.+)
              ]]>
            </Regex>
          </Expression>
        </Expressions>
      </Group>		
    </Groups>
  </ExpressionsToNotify>
  <appSettings>
	<add key="IPThreatApiKey" value=""/>	  
    <add key="FailedLoginAttemptsBeforeBan" value="25"/>
    <add key="BanTime" value="07:00:00:00"/>
    <add key="ResetFailedLoginCountForUnbannedIPAddresses" value="false"/>
    <add key="ClearBannedIPAddressesOnRestart" value="true"/>
    <add key="ClearFailedLoginsOnSuccessfulLogin" value="true"/>
    <add key="ProcessInternalIPAddresses" value="false"/>
    <add key="ExpireTime" value="00:01:00:00"/>
    <add key="CycleTime" value="00:00:10:00"/>
    <add key="MinimumTimeBetweenFailedLoginAttempts" value="00:00:00:01"/>
    <add key="MinimumTimeBetweenSuccessfulLoginAttempts" value="00:00:00:05"/>
    <add key="FirewallRulePrefix" value="AAAAIPBan_"/>
    <add key="Whitelist" value="2 ip's and 1 hostname"/>
    <add key="WhitelistRegex" value=""/>
    <add key="Blacklist" value="administrator"/>
    <add key="BlacklistRegex" value=""/>
    <add key="UserNameWhitelist" value="8 ccounts"/>
    <add key="UserNameWhitelistRegex" value=""/>
    <add key="UserNameWhitelistMinimumEditDistance" value="2"/>
    <add key="FailedLoginAttemptsBeforeBanUserNameWhitelist" value="50"/>
    <add key="ProcessToRunOnBan" value=""/>
    <add key="ProcessToRunOnUnban" value=""/>
    <add key="FirewallRules" value=""/>
    <add key="UseDefaultBannedIPAddressHandler" value="true"/>
    <add key="ExternalIPAddressUrl" value="https://checkip.amazonaws.com/"/>
    <add key="FirewallUriRules" value=""/>
    <add key="GetUrlUpdate" value=""/>
    <add key="GetUrlStart" value=""/>
    <add key="GetUrlStop" value=""/>
	<add key="GetUrlConfig" value="http://URL/ipban.config"/>
    <add key="IPBanPro_UrlWebSocket" value=""/>
  </appSettings>
</configuration>
`

**Comment by H3jake** [comment]
found the issue, a quick google search shows that when you download the file there is a hidden . at the betting of the file so i replaced
```
public async Task WriteConfigAsync(string xml)
        {
            // Ensure valid xml before writing the file
            XmlDocument doc = new();
            using (XmlReader xmlReader = XmlReader.Create(new StringReader(xml), new XmlReaderSettings { CheckCharacters = false }))
            {
                doc.Load(xmlReader);
            }
            await ConfigReaderWriter.WriteConfigAsync(xml);
        }
```
with this:
```
public async Task WriteConfigAsync(string xml)
        {
            // Ensure valid xml before writing the file
            XmlDocument doc = new();
            using (XmlReader xmlReader = XmlReader.Create(new StringReader(xml.remove(0,1)), new XmlReaderSettings { CheckCharacters = false }))
            {
                doc.Load(xmlReader);
            }
            await ConfigReaderWriter.WriteConfigAsync(xml);
        }
```
not the most correct or elegant solution but it works!

**Comment by H3jake** [comment]
https://stackoverflow.com/questions/17795167/xml-loaddata-data-at-the-root-level-is-invalid-line-1-position-1

also if you break before the doc.load then copy the xml into a notepad you should see the extra . added to the front

**Comment by jjxtra** [comment]
Hopefully resolved with da4a064591bdf194ab3f7a62a6e560f5bb76ff3c

**Comment by H3jake** [comment]
Hi Jeff,
I have just tried with your updated code but i get the same error.
